extensions/mssql/src/features.ts

```diff
@@ -43,7 +44,17 @@ export class AccountFeature implements StaticFeature {
 		let timeToLiveInSeconds = 10;
 		this.tokenCache = new DataItemCache(this.getToken, timeToLiveInSeconds);
 		this._client.onRequest(contracts.SecurityTokenRequest.type, async (request): Promise<contracts.RequestSecurityTokenResponse | undefined> => {
-			return this.tokenCache.getData(request);
+			return await this.tokenCache.getData(request);
+		});
+		this._client.onNotification(contracts.RefreshTokenNotification.type, async (request) => {
+			// Refresh token, then inform client the token has been updated. This is done as separate notification messages due to the synchronous processing nature of STS currently https://github.com/microsoft/azuredatastudio/issues/17179
+			let result = await this.refreshToken(request);
+			if (!result) {
+				void window.showErrorMessage(localizedConstants.tokenRefreshFailed);
+				console.log(`Token Refresh Failed ${request.toString()}`);
+				throw Error(localizedConstants.tokenRefreshFailed);
```

Comment:
in case of error, we should also notify STS right? otherwise it will be waiting forever.

Reply:
It would not be waiting, the token would just not get refreshed

Reply:
I know it is not technically waiting, but the current editor will never get appropriate intellisense because of this. we should let the user know how to get out of the bad state.

Reply:
+1 on this, and if an error happens it makes sense to at least clear out the URI from the token refresh map so it can try again next time it's requested. 

Showing an error isn't a bad idea, but be careful of spamming the user with repeated failed requests.

Reply:
You can consider other options too, like displaying a message telling the user that connected autocompletion is disabled due to token refresh error and then leaving it disconnected. If you do that though we should let the user re-enable it again, possibly just by clearing out the URI when "Refresh Intellisense Cache" command is ran

Reply:
the retry would happen pretty much immediately if the user is still actively working on the file. 
Ideally, depending on whether the error is recoverable, we should ask user whether to retry the token refresh or reconnect. up to Karl and Chris to decide the scope. but at least for now, we should let user know he needs to reconnect in order to get intellisense.